core/tests/puppeteer-acceptance-tests/puppeteer-testing-utilities/puppeteer-utils.js

```
@@ -80,6 +74,9 @@ module.exports = class puppeteerUtilities {
    */
   async signUpNewUser(userName, signInEmail) {
     await this.signInWithEmail(signInEmail);
+    if (this.page.url() === testConstants.URLs.LearnerDashboard) {
```

This seems wrong, why special case "learner dashboard" here in what should be a generic method? Instead, if the user gets logged in, then they are not "new" and the test should throw an error.

---

Yes @seanlip I totally agree ! This is just a patch i did so that the tests do not throw err. Yes, the structure here seems not fitting ! Because essentially we will maintain a single dev server without restarting it for each tests. It will run for the entireity of all tests. The lines which i added above is to address the issue of page redirections. When our user is already signed in, we will be straightaway redirected to the learner dashboard. In case we aren't registered with our email we will be asked to input a username and then proceed to the learner dashboard. I think we need to restructure this a bit and the naming for this function needs some tweaking too. cc @Shivkant-Chauhan

---

Yes @seanlip I totally agree ! This is just a patch i did so that the tests do not throw err. Yes, the structure here seems not fitting ! Because essentially we will maintain a single dev server without restarting it for each tests. It will run for the entireity of all tests. The lines which i added above is to address the issue of page redirections. When our user is already signed in, we will be straightaway redirected to the learner dashboard. In case we aren't registered with our email we will be asked to input a username and then proceed to the learner dashboard. I think we need to restructure this a bit and the naming for this function needs some tweaking too. cc @Shivkant-Chauhan

---

if in some cases, we will require to login an existing user, we can simply call that `signInWithEmail` function in the main tests only. but for all the current tests, there is no such need. also you need to have a fresh new datastore for each test seperately. If your tests are failing due to the error you explained above that means you donot have the store and the data cleared in your localhost includuing the users data.
by saying we need test independence means that if a user 'X' is created in test 1, then that user must not be present in some other test.

---

@Shivkant-Chauhan i understand we can opt this workflow when we are developing locally. But will this also work effectively with CI ? Essentially the github workflow we are working on will allow us to setup our devserver on a single runner and running the tests will be on a different runner. When working locally, as you commented we can clear our datastore by restarting the devserver , but replicating this same on CI; will it be efficient ? @seanlip @vojtechjelinek can you also please provide your view on this ?

---

@vojtechjelinek will tell you more about this.. we have discussed it earlier that for each test we will restart the google cloud server and firebase emulator server (instead of restarting the whole dev server), and this would be efficient as it would take upto 30-40 seconds

---

Note that there should be separate methods for signing up as a new user and logging in as an existing user, and the right one should be called at the right time.

As Shivkant says, it is also fine to clear and restart the server after each test. (Please also confirm that the server data still gets cleared if a test fails midway.)